### PR TITLE
MB-7933 Create segment for SAC/SDN when it is present

### PR DESCRIPTION
**Pull request by mandaem**
## Summary

This adds the SAC/SDN to EDI for each shipment service item in an additional FA2 segment with the ZZ qualifier `FA2*ZZ*GOESHERE`. (SAC and SDN live in the same field in MM)

If SAC/SDN is nil or an empty string, that segment should be omitted, if over 80 chars, it should show an error message.

Depending on` SACtype`, you'll see either the HHG (default) or NTS SAC/SDN.

## Verification Steps for the Author

These are to be checked by the author.

- [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
- [ ] Have the Jira acceptance criteria been met for this change?

## Verification Steps for Reviewers

These are to be checked by a reviewer.

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/run-storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/about/application-setup/milmove-local-client/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/about/development/testing)

### How to test

1. Go to the office app and Login as a TIO with gbloc KKFA
2. Search for move code "RDY4PY" and click in
3. Scroll down to payment requests, click Review, and approve 1+.
4. Refer to the [Acceptance Testing doc](https://transcom.github.io/mymove-docs/docs/backend/testing/acceptance-testing-syncada-edi-invoicing). 
5. In your terminal, if you have the json file from the doc in root dir, run `prime-api-client --insecure create-payment-request --filename ./payload.json | jq '.id,.paymentRequestNumber'` which should give you the 9 digit payment request ID for the following command
6. Run `make bin/prime-api-client bin/generate-payment-request-edi && generate-payment-request-edi --payment-request-number 6682-5285-1` replacing the end numbers with the payment request ID from previous
7. Confirm that the EDI prints in your terminal and does NOT include the new FA2 field, as hhg SAC's empty.
8. Go back into the move and add HHG and NTS SACs
9. Run 5 again and check that HHG SAC's in a new FA2 field

Screenshot 2023-05-18 at 12 17 00 PM

10. Make the hhg SACs 81+ chars and test 5 again. you should get an error msg in console noting the value's over 80 chars.
Screenshot 2023-05-19 at 9 06 10 AM

11. Go to the DB and change SACtype to NTS (see screenshot)

Screenshot 2023-05-19 at 9 20 54 AM

12. Run 5 again and check that NTS SAC, not HHG one, is in a new FA2 field
13. Make the nts SAC 81+ chars and test 5 again. you should get an error msg in console noting the value's over 80 chars.

### Frontend

- [ ] There are no aXe warnings for UI.
- [ ] This works in [Supported Browsers and their phone views](https://transcom.github.io/mymove-docs/docs/about/supported-browsers) (Chrome, Firefox, Edge).
- [ ] There are no new console errors in the browser devtools.
- [ ] There are no new console errors in the test output.
- [ ] If this PR adds a new component to Storybook, it ensures the component is fully responsive, OR if it is intentionally not, a wrapping div using the `officeApp` class or custom `min-width` styling is used to hide any states the would not be visible to the user.

### Backend

- [ ] Code follows the guidelines for [Logging](https://transcom.github.io/mymove-docs/docs/about/development/logging).
- [ ] The requirements listed in [Querying the Database Safely](https://transcom.github.io/mymove-docs/docs/backend/guides/golang-guide#querying-the-database-safely) have been satisfied.

#### Any new migrations/schema changes:

- [ ] Follows our guidelines for [Zero-Downtime Deploys](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#zero-downtime-migrations).
- [ ] Have been communicated to #g-database.
- [ ] Secure migrations have been tested following the instructions in our [docs](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#secure-migrations).

### CODE REVIEW

File: pkg/services/invoice/ghc_payment_request_invoice_generator.go [hunk 1]
```diff
@@ -555,16 +555,44 @@ func (g ghcPaymentRequestInvoiceGenerator) createLoaSegments(orders models.Order
...
 		tac = *orders.TAC
 	} else {
 		if orders.NtsTAC == nil || *orders.NtsTAC == "" {
-			return edisegment.FA1{}, edisegment.FA2{}, apperror.NewConflictError(orders.ID, "Invalid order. Must have an NTS TAC value")
+			return edisegment.FA1{}, nil, apperror.NewConflictError(orders.ID, "Invalid order. Must have an NTS TAC value")
 		}
 		tac = *orders.NtsTAC
 	}
 
+	// Get SAC or SDN from orders, using HHG one by default if it exists and NTS if not.
```
Comment (SirenaBorracha) [inline pkg/services/invoice/ghc_payment_request_invoice_generator.go:568]: I'm having trouble parsing what lines this comment applies to and there's enough ambiguity that I think anyone looking at this in the future would need to have a lot of context already in order to understand, would like to pair to talk through it

**Review by PDickeyTruss — APPROVED** [review]
Looks good, codes generated successfully in the EDI. Ari and I added a few comments that I think it would be helpful to briefly zoom with you to discuss on Monday (nothing blocking, just want to test our understanding).

File: pkg/services/invoice/ghc_payment_request_invoice_generator_test.go [hunk 2]
```diff
@@ -1419,6 +1412,64 @@ func (suite *GHCInvoiceSuite) TestTACs() {
...
+
+	suite.Run("shipment with no SAC type set", func() {
+		setupTestData()
+		mtoShipment.SACType = nil
+		suite.MustSave(&mtoShipment)
+
+		result, err := generator.Generate(suite.AppContextForTest(), paymentRequest, false)
+		suite.NoError(err)
+		suite.Equal(hhgSAC, result.ServiceItems[0].FA2s[1].FinancialInformationCode)
+	})
```
[the remark this answers is not in the document]
Comment (mandaem) [reply pkg/services/invoice/ghc_payment_request_invoice_generator_test.go:1424]: Thanks to you and @SirenaBorracha for noticing the comment is confusing. I'm going to revise it. HHG is the default.

File: pkg/services/invoice/ghc_payment_request_invoice_generator.go (hunk 1 above)
Comment (mandaem) [reply pkg/services/invoice/ghc_payment_request_invoice_generator.go:568]: I'm thinking I'll be asked to walk through this PR in WIP if that's ok? Also I'll be revising the comment so thanks for pointing this out.

File: pkg/services/invoice/ghc_payment_request_invoice_generator.go [hunk 3]
```diff
@@ -581,7 +609,19 @@ func (g ghcPaymentRequestInvoiceGenerator) createLoaSegments(orders models.Order
 		FinancialInformationCode:     tac,
 	}
 
-	return fa1, fa2, nil
+	fa2sac := edisegment.FA2{
```
Comment (reggieriser) [inline pkg/services/invoice/ghc_payment_request_invoice_generator.go:612]: **nitpick:** I think you could move this to after the `if` statement below as there's no need to create it unless we have a sac.

File: pkg/services/invoice/ghc_payment_request_invoice_generator_test.go [hunk 4]
```diff
@@ -802,9 +802,9 @@ func (suite *GHCInvoiceSuite) TestAllGenerateEdi() {
 			})
 
 			suite.Run("adds fa2 service item segment", func() {
-				fa2 := result.ServiceItems[segmentOffset].FA2
-				suite.Equal("TA", fa2.BreakdownStructureDetailCode)
-				suite.Equal(*paymentRequest.MoveTaskOrder.Orders.TAC, fa2.FinancialInformationCode)
+				fa2 := result.ServiceItems[segmentOffset].FA2s
+				suite.Equal("TA", fa2[0].BreakdownStructureDetailCode)
+				suite.Equal(*paymentRequest.MoveTaskOrder.Orders.TAC, fa2[0].FinancialInformationCode)
```
Comment (reggieriser) [inline pkg/services/invoice/ghc_payment_request_invoice_generator_test.go:807]: **question:** Now that we can have multiple FA2 segments, should we also assert on the slice length to ensure there's only one returned in this test?

File: pkg/services/invoice/ghc_payment_request_invoice_generator_test.go [hunk 5]
```diff
@@ -1358,7 +1362,7 @@ func (suite *GHCInvoiceSuite) TestTACs() {
 
 		result, err := generator.Generate(suite.AppContextForTest(), paymentRequest, false)
 		suite.NoError(err)
-		suite.Equal(hhgTAC, result.ServiceItems[0].FA2.FinancialInformationCode)
+		suite.Equal(hhgTAC, result.ServiceItems[0].FA2s[0].FinancialInformationCode)
```
Comment (reggieriser) [inline pkg/services/invoice/ghc_payment_request_invoice_generator_test.go:1365]: **suggestion:** Same thing with this one and other tests that look at the `FA2s` array -- probably makes sense to verify slice length is 1 or 2 depending on the scenario.  I don't think some of these tests will fail as they should if there was accidentally a second FA2s entry when there shouldn't be.

File: pkg/services/invoice/ghc_payment_request_invoice_generator.go (hunk 3 above)
Comment (mandaem) [reply pkg/services/invoice/ghc_payment_request_invoice_generator.go:612]: will do

File: pkg/services/invoice/ghc_payment_request_invoice_generator_test.go (hunk 4 above)
Comment (mandaem) [reply pkg/services/invoice/ghc_payment_request_invoice_generator_test.go:813]: I added checks for length and changed SAC population from BuildMove to local tests and added appropriate checks for that.

**Review by reggieriser — APPROVED** [review]

**Pull request merged**